umbrella.d.ts

```diff
@@ -1,207 +1,1484 @@
+     *
+     * @template NODE - The current iteration DOM element.
+     * @param {IterateCallback<NODE, Maybe<boolean>>} predicate - A function that tests each element in the set. It
+     *   should return `true` to keep the element, or `false` (or a falsy value) to remove it.
+     * @returns {this} - A new Umbrella instance containing the elements that passed the test.
+     */
+    filter(predicate: IterateCallback<NODE, Maybe<boolean>>): Umbrella<NODE>;
 
-    first(): Element | false;
+    /**
+     * Remove all the nodes that doesn't match the criteria.
+     *
+     * [Documentation](https://umbrellajs.com/documentation#filter)
+     *
+     * @example
+     *   u('div').filter(u('a'));
+     *
+     * @template NODE - Defines the type of DOM elements that the Umbrella instance manipulates.
+     * @param {Umbrella} filter - Instance of Umbrella with the elements to keep (the intersection will be kept).
+     * @returns {this} - A new Umbrella instance containing the filtered set of elements.
+     */
+    filter(filter: Umbrella): Umbrella<NODE>;
 
-    data: typeof u.attr;
+    /**
+     * Remove all the nodes that doesn't match the criteria.
+     *
+     * [Documentation](https://umbrellajs.com/documentation#filter)
+     *
+     * @example
+     *   // Get only the active links
+     *   const activeLinks = u('a').filter('.active');
+     *
+     *   // Using null to get a copy Umbrella instance
+     *   const copy = u('div').filter(null);
+     *
+     * @template NODE - Defines the type of DOM elements that the Umbrella instance manipulates.
+     * @param {Maybe<string>} filter - A CSS selector string used to filter the matched elements. `null` or `undefined`
+     *   replaces with * selector.
+     * @returns {this} - A new Umbrella instance containing the filtered set of elements.
+     */
+    filter(filter: Maybe<string>): Umbrella<NODE>;
 
-    each(callback: (node: Element, i: number) => void): this;
+    /**
+     * Searches for elements that satisfy a condition defined by the provided predicate function.
+     *
+     * [Documentation](https://umbrellajs.com/documentation#find)
+     *
+     * @example
+     *   // Get the required fields within a submitting form:
+     *   u('form').on('submit', function () {
+     *     const required = u(this).find(node => node.hasClass('required'));
+     *   });
+     *
+     * @template NODE - The current iteration DOM element.
+     * @template RES_NODE - Specifies the type of the resulting elements found.
+     * @param {IterateCallback<NODE, Maybe<boolean>>} predicate - A function that tests each element in the set.
+     *   Elements for which the function returns `true` are included in the resulting Umbrella instance.
+     * @returns {Umbrella<RES_NODE>} - An Umbrella instance containing the elements that match the predicate.
+     */
+    find<RES_NODE extends Element = Element>(predicate: IterateCallback<NODE, Maybe<boolean>>): Umbrella<RES_NODE>;
```

Comment:
You seem right, `.find()` with a callback seems not to work properly, we definitely need [a test](https://github.com/franciscop/umbrella/blob/master/src/plugins/find/test.js) for that use-case! But following `.find()` with a callback:

```js
// The method
u.prototype.find = function (selector) {
  return this.map(function (node) {
    return u(selector || '*', node);
  });
};

// If selector is a function, it will fall here on the constructor:
this.nodes = this.slice(parameter);

// slice behaves like this:
u.prototype.slice = function (pseudo) {
  // Check that it's not a valid object
  if (!pseudo ||
      pseudo.length === 0 ||
      typeof pseudo === 'string' ||
      pseudo.toString() === '[object Function]') return [];
```

We def see that in that case it returns an empty array instead of a list of filtered nodes, good catch! A new legit bug after these many years and usage. I guess the use-case is _very_ rare (and "kinda" undocumented, at least no examples).

Or, we could just mention that is not supported. But I do think it's a bug/should be fixed. The docs _do_ say that "filter: a string containing a selector" and then points to `.filter()` for more explanation, which is wrong.